main/src/library/Graph.flix

```diff
@@ -0,0 +1,92 @@
+/*
+ *  Copyright 2022 Nina Andrup Pedersen
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+namespace Graph {
+
+    ///
+    /// Returns the transitive closure of the graph `g`.
+    ///
+    pub def closure(g: m[(t, t)]): Set[(t, t)] & Impure with Foldable[m], Boxable[t] =
+        let edges = inject g into Edge;
+
+        let reachability = #{
+            // All nodes can reach themselves.
+            Reachable(n1, n1) :- Edge(n1, _).
+            Reachable(n2, n2) :- Edge(_, n2).
+            // `n1` can reach `n2` by using an edge.
+            Reachable(n1, n2) :- Edge(n1, n2).
+            // if `n1` can reach `m` and `m` can reach `n2` then `n1` can
+            // also reach `n2`. This adds node pairs to the relational
+            // that is reachable using any number of nodes.
+            Reachable(n1, n2) :- Reachable(n1, m), Reachable(m, n2).
+        };
+        let x = query edges, reachability select (src, dst) from Reachable(src, dst);
+        Array.toSet(x)
```

Comment:
The type of `x` is `Array[(t, t), False]`, which makes the functions `Impure`. I guess that is not what we want, but I was not sure how to wrap it in a region. This does not work:
```
region r {
    let x = (query edges, reachability select (src, dst) from Reachable(src, dst)) @ r;
    ...
}
```

Reply:
Currently query returns array (which is impure). That will change, but in the mean time, I suggest to use a cast on the  whole thing.